rest_server/IOMgrConnection.go

```diff
@@ -0,0 +1,393 @@
+	IOMgrConn.RequestID++
+	var databaseEngineRestRequest SiodbIomgrProtocol.DatabaseEngineRestRequest
+	databaseEngineRestRequest.RequestId = requestId
+	databaseEngineRestRequest.Verb = restVerb
+	databaseEngineRestRequest.ObjectType = objectType
+	databaseEngineRestRequest.UserName = userName
+	databaseEngineRestRequest.Token = token
+
+	if len(objectName) > 0 {
+		databaseEngineRestRequest.ObjectName = objectName
+	}
+	if objectId > 0 {
+		databaseEngineRestRequest.ObjectId = objectId
+	}
+	siodbLoggerPool.Debug("databaseEngineRestRequest: %v", databaseEngineRestRequest)
+	siodbLoggerPool.Debug("IOMgrConn.RequestID: %v", requestId)
+	siodbLoggerPool.Debug("IOMgrConn.RequestID++: %v", IOMgrConn.RequestID)
+
+	if _, err := IOMgrConn.writeMessage(DATABASEENGINERESTREQUEST, &databaseEngineRestRequest); err != nil {
+		return requestId, fmt.Errorf("Unable to write message to IOMgr: %v", err)
+	}
+
+	return requestId, nil
+}
+
+func (IOMgrConn *IOMgrConnection) readIOMgrResponse(requestID uint64) (err error) {
+
+	var databaseEngineResponse SiodbIomgrProtocol.DatabaseEngineResponse
+
+	if _, err := IOMgrConn.readMessage(DATABASEENGINERESPONSE, &databaseEngineResponse); err != nil {
+		return fmt.Errorf("Unable to read response from IOMgr: %v", err)
+	}
+	siodbLoggerPool.Debug("databaseEngineResponse: %v", databaseEngineResponse)
+	siodbLoggerPool.Debug("databaseEngineResponse.RequestId: %v", databaseEngineResponse.RequestId)
+	siodbLoggerPool.Debug("IOMgrConn.RequestID: %v", IOMgrConn.RequestID)
+
+	if databaseEngineResponse.RequestId != requestID {
+		return fmt.Errorf("request IDs mismatch")
+	}
+
+	if len(databaseEngineResponse.Message) > 0 {
+		return fmt.Errorf("code: %v, message: %v", databaseEngineResponse.Message[0].GetStatusCode(), databaseEngineResponse.Message[0].GetText())
+	}
+
+	return nil
+}
+
+func (IOMgrConn *IOMgrConnection) readChunkedJSON(c *gin.Context) (err error) {
+
+	// Stream through HTTP > 1.1
+	// https://stackoverflow.com/questions/29486086/how-http2-http1-1-proxy-handle-the-transfer-encoding
+
+	var IOMgrChunkSize uint32
+	var IOMgrbuffSize uint32
+	var IOMgrBytesRead int
+	var IOMgrbuffRemain []byte
+	var JSONChunk string
+	var JSONChunkBuff string
+
+	for {
+
+		// Get Current IOMgr chunk Size
+		if _, IOMgrChunkSize, err = IOMgrConn.readVarint32(); err != nil {
+			return fmt.Errorf("Not able to read chunk size from IOMgr: %v", err)
+		}
+		siodbLoggerPool.Debug("IOMgrChunkSize: %v", IOMgrChunkSize)
+
+		if IOMgrChunkSize == 0 {
+			break
+		}
+
+		for IOMgrChunkSize != 0 {
+			if IOMgrChunkSize < IOMgrChunkMaxBufferedSize {
+				IOMgrbuffSize = IOMgrChunkSize
+				IOMgrChunkSize = 0
+			} else if IOMgrChunkSize >= IOMgrChunkMaxBufferedSize {
+				IOMgrbuffSize = IOMgrChunkMaxBufferedSize
+				IOMgrChunkSize = IOMgrChunkSize - IOMgrChunkMaxBufferedSize
+			}
+			siodbLoggerPool.Debug("IOMgrbuffSize: %v", IOMgrbuffSize)
+
+			IOMgrbuff := make([]byte, uint32(len(IOMgrbuffRemain))+IOMgrbuffSize)
```

Comment:
why allocate this buffer each time in the loop? did you know that memory allocations are one of most costly operations that affect performace? It can be allocated once with size  (IOMgrChunkMaxBufferedSize  + 4) (4 is max utf8 character sequence size).